packages/cli/src/util/get-latest-version/get-latest-worker.js

```diff
@@ -14,36 +14,101 @@ const fetch = require('node-fetch');
 const fs = require('fs-extra');
 const path = require('path');
 const { Agent: HttpsAgent } = require('https');
+const { bold, gray, red } = require('chalk');
+const { format, inspect } = require('util');
 
-process.on('unhandledRejection', err => {
-  if (process.connected) {
-    console.error('Exiting update worker due to error:');
-    console.error(err);
+/**
+ * An simple output helper which accumulates error and debug log messages in
+ * memory for potential persistance to disk while immediately outputting errors
+ * and debug messages, when the `--debug` flag is set, to `stderr`.
+ */
+class WorkerOutput {
+  debugLog = [];
+  logFile = null;
+
+  constructor({ debug = true }) {
+    this.debugOutputEnabled = debug;
+  }
+
+  debug(...args) {
+    this.print('debug', args);
+  }
+
+  error(...args) {
+    this.print('error', args);
+  }
+
+  print(type, args) {
+    const str = format(
+      ...args.map(s => (typeof s === 'string' ? s : inspect(s)))
+    );
+    this.debugLog.push(`[${new Date().toISOString()}] [${type}] ${str}`);
+    if (type === 'debug' && this.debugOutputEnabled) {
+      console.error(
+        `${gray('>')} ${bold('[debug]')} ${gray(
+          `[${new Date().toISOString()}]`
+        )} ${str}`
+      );
+    } else if (type === 'error') {
+      console.error(`${red(`Error:`)} ${str}`);
+    }
   }
+
+  setLogFile(file) {
+    if (this.logFile === null) {
+      process.on('exit', () => {
+        if (this.debugLog.length) {
+          fs.outputFileSync(this.logFile, this.debugLog.join('\n'));
+        }
+      });
+    }
+    this.logFile = file;
+  }
+}
+
+const output = new WorkerOutput({
+  debug: process.argv.includes('--debug') || !process.connected,
```

Comment:
**suggestion (non-blocking):** check `process.connected` before debug logging

`process.connected` will always be true here, right? The value changes later after the IPC message is received. Should this value be checked right before debug logging?

`debugOutputEnabled` could become a getter that checks the underlying value as well as `process.connected`. I wouldn't be a fan of the running instance constantly checking global values, but it would work. Wrapping the access to `process.connected` in its own getter would at least allow a test to override the global access behavior.

Example:

```
class WorkerOutput {
  get processConnected() {
    return process.connected;
  }

  get debugOutputEnabled() {
    return this._debugOutputEnabled || !this.processConnected;
  }

  // ...
}
```

Reply:
There are 2 ways the worker script can be invoked:

1. Manually run `node /path/to/get-latest-worker.js`. For this we force debug logging so that we see any unhandled rejection errors, of which should never happen. Manually running the script should just print `"No IPC bridge detected, exiting"`. If we don't care about unhandled rejections, which should never occur, we could simply remove the `process.connected` all together.

2. The worker is spawned by the CLI, `process.connected` will be `true` and `debug` will be `false` by default. If `--debug` is set, then the debug log is printed to `stderr`.

We don't need to overcomplicate the debug logging with checking `process.connected` because the debug logger doesn't care if the worker process is connected or not. The `stderr` file descriptor is still valid even after the processes disconnect.

Reply:
I was reading too much intent behind the check. This sounds good to me.